history_and_modals.md

```diff
@@ -1,158 +1,357 @@
+
+Additionally, note that the inability to prevent closing of `<dialog>`s via <kbd>Esc</kbd> makes them hard to use for some web developers, and requires them to roll their own custom dialog modals. _TODO: evidence of this would be good._
+
+On mobile platforms, getting the right behavior is significantly harder. First, platform-specific code is required, to do nothing on iOS and to capture the back button on Android. Next, capturing the back button on Android requires manipulating the history stack using the [history API](https://developer.mozilla.org/en-US/docs/Web/API/History_API). This is a poor fit for several reasons:
+
+- This UI state is non-navigational, i.e., the URL doesn't and shouldn't change when opening a modal. When the page is reloaded or shared, web developers generally don't want to automatically re-open the same modal.
+
+- It's very hard to determine when the modal state is popped out of history, because both moving forward and backward through history emit the same `popstate` event.
+
+- The `history.state` API, used to store the modal state, is rather fragile: user-initiated fragment navigation, as well as other calls to `history.pushState()` and `history.replaceState()`, can override it.
+
+- History navigation is not cancelable. Implementing "Are you sure you want to close this dialog without saving?" requires letting the history navigation complete, then potentially re-establishing the modal state if the user declines to close the modal.
+
+- A shared component that attempts to use the history API to implement these techniques can easily corrupt a web application's router.
+
+## Proposal
+
+The proposal is to introduce a new API, the `ModalCloseWatcher` class, which has the following basic API:
+
+```js
+// Constructor only works from within transient user activation;
+// otherwise, it throws a "NotAllowedError" DOMException.
+//
+// Only one such construction can be done per transient user activation,
+// but constructing does not consume user activation.
+const watcher = new ModalCloseWatcher();
+
+// This fires when the user sends a close signal.
+watcher.onclose = () => {
+  myModal.close();
+};
+
+// You should destroy watchers which are no longer needed, e.g. if the
+// modal closes normally. This will prevent future events on this watcher.
+myModalCloseButton.onclick = () => {
+  watcher.destroy();
+};
+```
+
+If more than one `ModalCloseWatcher` is active at a given time, then only the most-recently-constructed one gets events delivered to it. A watcher becomes inactive after a `close` event is delivered, or the watcher is explicitly `destroy()`ed.
```

Comment:
Would we expect focus to have something to do with this choice?

Imagine you have two date pickers open. A user switches focus back and forth between them, editing the dates. Now you send the cancel signal. Should it be sent to whichever date picker was opened last, or should it be the one that you've last modified.

Do we want to worry about cases with more than one modal control visible at all?

Reply:
Interesting idea. I'd been operating under the assumption that, by definition, only the most recent modal would be considered topmost. I'm not aware of UI patterns that allow a different setup, e.g. all datepickers I'm aware of close when clicking elsewhere on the page (including trying to open another datepicker).